twingate/internal/client/resource.go

```diff
@@ -320,3 +328,113 @@ func (client *Client) readResourcesByNameAfter(ctx context.Context, variables ma
 
 	return &response.PaginatedResource, nil
 }
+
+func (client *Client) DeleteResourceServiceAccounts(ctx context.Context, resourceID string, deleteServiceAccountIDs []string) error {
+	if len(deleteServiceAccountIDs) == 0 {
+		return nil
+	}
+
+	if resourceID == "" {
+		return NewAPIError(ErrGraphqlIDIsEmpty, operationUpdate, resourceResourceName)
+	}
+
+	resourcesToDelete := []string{resourceID}
+
+	for _, serviceAccountID := range deleteServiceAccountIDs {
+		if err := client.UpdateServiceAccountRemoveResources(ctx, serviceAccountID, resourcesToDelete); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (client *Client) AddResourceGroups(ctx context.Context, resource *model.Resource) error {
+	if len(resource.Groups) == 0 {
+		return nil
+	}
+
+	if resource.ID == "" {
+		return NewAPIError(ErrGraphqlIDIsEmpty, operationUpdate, resourceResourceName)
+	}
+
+	variables := newVars(
+		gqlID(resource.ID),
+		gqlIDs(resource.Groups, "groupIds"),
+	)
+
+	response := query.AddResourceGroups{}
+
+	err := client.GraphqlClient.NamedMutate(ctx, "updateResource", &response, variables)
+	if err != nil {
+		return NewAPIErrorWithID(err, operationUpdate, resourceResourceName, resource.ID)
+	}
+
+	if !response.Ok {
+		return NewAPIErrorWithID(NewMutationError(response.Error), operationUpdate, resourceResourceName, resource.ID)
+	}
+
+	return nil
+}
+
+func (client *Client) DeleteResourceGroups(ctx context.Context, resourceID string, deleteGroupIDs []string) error {
+	if len(deleteGroupIDs) == 0 {
+		return nil
+	}
+
+	if resourceID == "" {
+		return NewAPIError(ErrGraphqlIDIsEmpty, operationUpdate, resourceResourceName)
+	}
+
+	response := query.UpdateResourceRemoveGroups{}
+	variables := newVars(
+		gqlID(resourceID),
+		gqlIDs(deleteGroupIDs, "removedGroupIds"),
+	)
+
+	err := client.GraphqlClient.NamedMutate(ctx, "updateResource", &response, variables)
+	if err != nil {
+		return NewAPIErrorWithID(err, operationUpdate, resourceResourceName, resourceID)
+	}
+
+	if !response.Ok {
+		return NewAPIErrorWithID(NewMutationError(response.Error), operationUpdate, resourceResourceName, resourceID)
+	}
+
+	if response.Entity == nil {
+		return NewAPIErrorWithID(ErrGraphqlResultIsEmpty, operationUpdate, resourceResourceName, resourceID)
+	}
+
+	return nil
+}
+
+func (client *Client) ReadResourceServiceAccounts(ctx context.Context, resourceID string) ([]string, error) {
+	serviceAccounts, err := client.ReadServiceAccounts(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccountIDs := make(map[string]bool)
```

Comment:
Why is it a map you need to convert back to list in line 425 and not just a list you append to?

Reply:
Are you trying to de-dupe duplicate service account ids here? why?

Reply:
yes, I think here we don't need to do de-duplication